### PR TITLE
Initial SHMAM implementation

### PR DESCRIPTION
**Pull request by charlesarcher**

### CODE REVIEW

**Comment by raffenet** [comment]
test:jenkins/ch4/all

**Comment by raffenet** [comment]
test:jenkins/ch4/all

**Comment by raffenet** [comment]
@charlesarcher build issues: https://jenkins.mpich.org/job/mpich-review-ch4-ofi/244/compiler=gnu,jenkins_configure=default,label=ib64,ofi_prov=sockets/console

**Comment by raffenet** [comment]
test:jenkins/ch4/all

**Comment by raffenet** [comment]
test:jenkins/ch4/all

**Comment by raffenet** [comment]
test:jenkins/ch4/all

**Comment by raffenet** [comment]
test:jenkins/ch4/ucx

**Comment by raffenet** [comment]
test:jenkins/ch4/all

**Comment by raffenet** [comment]
test:jenkins/ch4/ucx

**Comment by raffenet** [comment]
All the build issues have been addressed. Removing the commit to make shmam the default and running tests again to catch any potential regressions to the old code.

test:jenkins/ch4/all

**Comment by raffenet** [comment]
test:jenkins/ch4/all

File: src/mpid/ch4/include/mpidpre.h [hunk 1]
```diff
@@ -159,6 +159,8 @@ typedef struct MPIDI_CH4U_req_ext_t {
 typedef struct MPIDI_CH4U_req_t {
     union {
     MPIDI_NM_REQUEST_AM_DECL} netmod_am;
+    union {
+    MPIDI_SHM_REQUEST_AM_DECL} shm_am;
```
Comment (pavanbalaji) [inline src/mpid/ch4/include/mpidpre.h]: Shouldn't the netmod_am and shm_am be inside a single union?  We'll only use one at a time (whichever is the transport).

File: src/mpid/ch4/shm/shmam/eager/fbox/fbox_init.h [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+#include "fbox_types.h"
+
+extern MPIDI_SHMAM_eager_fbox_control_t MPIDI_SHMAM_eager_fbox_control_global;
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_SHMAM_eager_init
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_SHMAM_eager_init(int rank, int grank, int num_local)
```
Comment (pavanbalaji) [inline src/mpid/ch4/shm/shmam/eager/fbox/fbox_init.h]: This and several other functions are missing the state declarations.

File: src/mpid/ch4/shm/shmam/eager/fbox/fbox_init.h [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    /* Allocate table of pointers to fastboxes */
+
+    MPIR_CHKPMEM_MALLOC(MPIDI_SHMAM_eager_fbox_control_global.mailboxes.in,
+                        MPIDI_SHMAM_fastbox_t **, num_local * sizeof(MPIDI_SHMAM_fastbox_t *),
+                        mpi_errno, "fastboxes");
+    MPIR_CHKPMEM_MALLOC(MPIDI_SHMAM_eager_fbox_control_global.mailboxes.out,
+                        MPIDI_SHMAM_fastbox_t **, num_local * sizeof(MPIDI_SHMAM_fastbox_t *),
+                        mpi_errno, "fastboxes");
+
+#define MPIDI_SHMAM_MAILBOX_INDEX(sender, receiver) ((num_local) * (sender) + (receiver))
```
Comment (pavanbalaji) [inline src/mpid/ch4/shm/shmam/eager/fbox/fbox_init.h]: Defining something in between a function is rather gross.  Can this be moved to the top?

File: src/mpid/ch4/shm/shmam/eager/fbox/fbox_init.h [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    MPIR_CHKPMEM_REAP();
+    goto fn_exit;
+    /* --END ERROR HANDLING-- */
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_SHMAM_eager_finalize
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_SHMAM_eager_finalize()
```
Comment (pavanbalaji) [inline src/mpid/ch4/shm/shmam/eager/fbox/fbox_init.h]: The namespace is rather weird here.  The API exposed to the glue layer would be MPIDI_SHMAM_.  Is "eager_finalize" meant to be exposed to the glue layer or internal?  If it is internal, shouldn't the naming be MPIDI_SHMAMI_?

File: src/mpid/ch4/shm/shmam/eager/include/shmam_eager_impl.h [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+ */
+#ifndef SHM_SHMAM_EAGER_IMPL_H_INCLUDED
+#define SHM_SHMAM_EAGER_IMPL_H_INCLUDED
+
+#ifndef SHMAM_EAGER_DIRECT
+#ifndef SHMAM_EAGER_DISABLE_INLINES
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_SHMAM_eager_init(int rank, int grank, int size)
+{
+    return MPIDI_SHMAM_eager_func->init(rank, grank, size);
```
Comment (pavanbalaji) [inline src/mpid/ch4/shm/shmam/eager/include/shmam_eager_impl.h]: I'm not sure why we need these function pointers in the shmmod layer at all.  All possible shmmods are always compiled in (they have different namespaces).  This is different from the netmod architecture, where all netmods expose the same API.

**Comment by wesbland** [comment]
This PR can probably be closed and @ddurnov can open a new PR when he is ready (based on the new glue code from @pavanbalaji).

**Pull request closed without merging**